File: app/src/main/java/com/yeputra/moviecatalogue/base/BaseRecyclerViewAdapter.java
package com.yeputra.moviecatalogue.base;

import android.widget.Filter;
import android.widget.Filterable;

import java.util.ArrayList;
import java.util.List;

import androidx.recyclerview.widget.RecyclerView;


public abstract class BaseRecyclerViewAdapter<VH extends RecyclerView.ViewHolder, T>
        extends RecyclerView.Adapter<VH>
        implements Filterable, FilterResultListener<T> {

    private List<T> items;

    protected boolean isSearching = false;

    protected abstract void onBindViewHolder(VH holder, T item, int position);

    public BaseRecyclerViewAdapter() {
        this.items = new ArrayList<>();
    }


    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            public Filter.FilterResults performFiltering(CharSequence constraint) {
                isSearching = true;
                String charString = constraint.toString();
                Filter.FilterResults filterResults = new Filter.FilterResults();

                items = onFilterResult(charString);

                filterResults.values = items;
                filterResults.count = getItemCount();
                return filterResults;
            }

            @Override
            public void publishResults(CharSequence constraint, Filter.FilterResults results) {
                items = (List<T>) results.values;
                notifyDataSetChanged();
                isSearching = false;
            }
        };
    }

    @Override
    public List<T> onFilterResult(String query) {
        return this.items;
    }

    @Override
    public int getItemCount() {
        return this.items == null ? 0 : this.items.size();
    }

    @Override
    public void onBindViewHolder(VH holder, int position){
        onBindViewHolder(holder, this.items.get(position), position);
    }

    public T getItem(int position) { return this.items.get(position); }

    public void addItem(T item) {
        this.items.add(item);
        this.notifyDataSetChanged();
    }

    public void appendItem(List<T> items){
        this.items.addAll(items);
        this.notifyDataSetChanged();
    }

    public void replaceItem(List<T> items){
        this.items.clear();
        this.appendItem(items);
        this.notifyDataSetChanged();
    }
}

File: app/src/main/java/com/yeputra/moviecatalogue/base/BaseRecyclerViewAdapter.kt
package com.yeputra.moviecatalogue.base

import android.widget.Filter
import android.widget.Filterable
import androidx.recyclerview.widget.RecyclerView


/**
 * Created by yovi.putra
 *    on 14/Feb/2019 15:36
 * Company SIEMO - PT. Multipolar Technology, Tbk
 */
interface FilteringListener <T>{
    fun onFiltering(query: String?): MutableList<T>
}

abstract class BaseRecyclerViewAdapter <VH: RecyclerView.ViewHolder,T>
    : RecyclerView.Adapter<VH>(), Filterable, FilteringListener<T> {

    private var items: MutableList<T>
    private var tempItems: MutableList<T>

    protected var isSearching: Boolean = false

    abstract fun onBindViewHolder(holder: VH, item: T, position: Int)

    init {
        items = mutableListOf()
        tempItems = mutableListOf()
    }

    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(constraint: CharSequence): FilterResults {
                isSearching = true
                val charString = constraint.toString()
                val filterResults = FilterResults()
                val filterResult = onFiltering(charString)

                filterResults.values = filterResult
                filterResults.count = filterResult.size
                return filterResults
            }

            override fun publishResults(constraint: CharSequence, results: FilterResults) {
                @Suppress("UNCHECKED_CAST")
                items = results.values as MutableList<T>
                notifyDataSetChanged()
                isSearching = false
            }
        }
    }

    override fun onFiltering(query: String?): MutableList<T> {
        return this.tempItems
    }

    override fun getItemCount(): Int {
        return this.items.size
    }

    override fun onBindViewHolder(holder: VH, position: Int){
        onBindViewHolder(holder, this.items[position], position)
    }

    fun getItem(position: Int) : T? = items[position]

    private fun addItem(items: MutableList<T>){
        this.items.addAll(items)
        this.tempItems.addAll(items)
        this.notifyDataSetChanged()
    }

    private fun clear(){
        this.items.clear()
    }

    fun setItem(item: MutableList<T>) {
        clear()
        addItem(item)
    }
}